include/nbl/builtin/hlsl/bxdf/ndf/ggx.hlsl

```diff
@@ -16,93 +16,150 @@ namespace bxdf
     }
 
-    scalar_type correlated_wo_numerator(scalar_type a2, scalar_type NdotV, scalar_type NdotV2, scalar_type NdotL, scalar_type NdotL2)
+    // without numerator, numerator is 2 * NdotV * NdotL, we factor out 4 * NdotV * NdotL, hence 0.5
+    template<class Query, class LS, class Interaction NBL_FUNC_REQUIRES(ggx_concepts::G2XQuery<Query> && LightSample<LS> && surface_interactions::Isotropic<Interaction>)
+    scalar_type correlated(NBL_CONST_REF_ARG(Query) query, NBL_CONST_REF_ARG(LS) _sample, NBL_CONST_REF_ARG(Interaction) interaction)
     {
-        scalar_type one_minus_a2 = scalar_type(1.0) - a2;
-        scalar_type Vterm = NdotL * devsh_part(NdotV2, a2, one_minus_a2);
-        scalar_type Lterm = NdotV * devsh_part(NdotL2, a2, one_minus_a2);
+        BxDFClampMode _clamp = query.getClampMode();
+        assert(_clamp != BxDFClampMode::BCM_NONE);
+
+        scalar_type Vterm = _sample.getNdotL(_clamp) * query.getDevshV();
+        scalar_type Lterm = interaction.getNdotV(_clamp) * query.getDevshL();
         return scalar_type(0.5) / (Vterm + Lterm);
     }
 
-    scalar_type G2_over_G1(scalar_type a2, bool transmitted, scalar_type NdotV, scalar_type NdotV2, scalar_type NdotL, scalar_type NdotL2)
+    template<class Query, class LS, class Interaction, class MicrofacetCache NBL_FUNC_REQUIRES(ggx_concepts::G2XQuery<Query> && LightSample<LS> && surface_interactions::Isotropic<Interaction> && ReadableIsotropicMicrofacetCache<MicrofacetCache>)
+    scalar_type G2_over_G1(NBL_CONST_REF_ARG(Query) query, NBL_CONST_REF_ARG(LS) _sample, NBL_CONST_REF_ARG(Interaction) interaction, NBL_CONST_REF_ARG(MicrofacetCache) cache)
     {
+        BxDFClampMode _clamp = query.getClampMode();
+        assert(_clamp != BxDFClampMode::BCM_NONE);
+
         scalar_type G2_over_G1;
-        if (transmitted)
+        scalar_type NdotV = interaction.getNdotV(_clamp);
+        scalar_type NdotL = _sample.getNdotL(_clamp);
+        scalar_type devsh_v = query.getDevshV();
+        scalar_type devsh_l = query.getDevshL();
+        if (cache.isTransmission())
         {
-            scalar_type one_minus_a2 = scalar_type(1.0) - a2;
-            scalar_type devsh_v = devsh_part(NdotV2, a2, one_minus_a2);
-            scalar_type L_v = scalar_type(0.5) * (devsh_v / NdotV - scalar_type(1.0));
-            scalar_type L_l = scalar_type(0.5) * (devsh_part(NdotL2, a2, one_minus_a2) / NdotL - scalar_type(1.0));
-            G2_over_G1 = hlsl::beta<scalar_type>(scalar_type(1.0) + L_l, scalar_type(1.0) + L_v);
-            G2_over_G1 *= scalar_type(1.0) + L_v;
+            if (NdotV < 1e-7 || NdotL < 1e-7)
+                return 0.0;
+            scalar_type onePlusLambda_V = scalar_type(0.5) * (devsh_v / NdotV + scalar_type(1.0));
+            scalar_type onePlusLambda_L = scalar_type(0.5) * (devsh_l / NdotL + scalar_type(1.0));
+            G2_over_G1 = bxdf::beta<scalar_type>(onePlusLambda_L, onePlusLambda_V) * onePlusLambda_V;
```

Comment:
`devsh_x/NdotX` will go to NaN because of `0/0`  (mathematically they should go to INF)

here use `bxdf::beta::__call_wo_checks` do not check `onePlusLambda_L` and `onePlusLambda_V` for being large because we'll never call it with large arguments from here

Reply:
continues https://github.com/Devsh-Graphics-Programming/Nabla/pull/899#discussion_r2216355519